### DISCUSSION TITLE
Query abstraction

### DISCUSSION THREAD

**Discussion by puschie286** [abpframework/abp#11513] [Show and tell]
Hi,
we encounter some problem while abstracting parts of queries. some systems use the same 'base' query and put a few coditions on top of that. after a lot of different approches ( which mostly failed because of linq to sql translation problems ) we found [LinqKit](https://github.com/scottksmith95/LINQKit) and were able to abstract just the specific parts of the query and combine multiple expressions to create the final query with ease.

As example: imagine you have a multiple entites with a date range and each entity define some specific data.
Now you want to write a base domain service that add the date range check to the specific entity query. 

BaseDomainService:
```
protected abstract Expression<Func<Entity, bool>> GetEntityExpression( Parameter param );

public async Task<List<Entity>> GetData( Parameter param, CancellationToken cancel = default )
{
   return await AsyncExecuter.ToListAsync(
      ( await Repository.GetQueryableAsync() ).AsExpandable().Where( entry =>
         GetEntityExpression( param ).Invoke( entry ) &&
         entry.Start <= param.date &&
         entry.End >= param.date ),
      cancel
   );
}
```
Note: 'Start' and 'End' describe the validity of the data and 'date' is the requested DateTime the data should be valid for
( You need to add LinqKit.Core to your Domain project for this )

Im new to the usage of expression with linq to sql, so i guess its possible to do this without LinqKit but it makes the usage pretty clean and straight forward.

Hope this helps some of you :)

**Comment by realLiangshiwei** [comment]
How about specification pattern: https://docs.abp.io/en/abp/latest/Specifications

**Reply by puschie286** [reply]
they are nice to abstract fixed conditions. but how do you force derived class to implement specific condition checks ? 
i dont see a way to use them as interface / abstract method ( its also more complicated if you need multiple ).
